### DISCUSSION TITLE
How did you convert the model to ONNX?

### DISCUSSION THREAD

**Discussion by TechInterMezzo** [snakers4/silero-vad#531] [Q&A]
This is out of curiosity and to learn from others. I don't need your source code.

How did you convert your model from PyTorch to ONNX?
Did you use torch.onnx.export or torch.onnx.dynamo_export?
Did you have to adjust your model code to make it exportable to ONNX?

**Comment by hx0810** [comment]
I have the same problem. When converting the onnx model, an error "Expected node type 'onnx::Constant', got 'onnx::Cast'" appears.
![1727339419459](https://github.com/user-attachments/assets/e42dcd7a-1fe7-45ab-96a8-5f99a04ef144)

**Comment by MisteryX** [comment]
Converting of the model now a little bit tricky because of the progress of torch project. 
Torchscript in the model now became invalid, so before exporting it, you should modify model file **silero_vad.jit (zip archive) -> VADr_v5/code/__torch__/vad/model/vad_annotator.py:58**
Replace invalid (in modern versions of torch) string: 
`_9 = torch.__not__(bool(torch.len(_context)))`
with
`_9 = torch.eq(torch.len(_context), 0)`
Build new model archive file .jit using ZIP arch.

Now you can export JIT model to ONNX file, using code:
`torch.onnx.export(
            model,
            (torch.randn(1, 512), torch.tensor([[16000]], dtype=torch.int64)),
            f=file_name,
            input_names=['input', 'sr'],
            output_names=['output'],
            dynamic_axes={
                'input': {0: 'batch size'},
                'output': {0: 'batch size'}
            },
            verbose=True,
            verify=True,
        )`
But you will get the Merged version of the model with specified inputs\outputs. How to obtain model version with state input and output params is the question for the model developers.

**Comment by fanlu** [comment]
@adamnsandle Could you share your onnx export script please? I appreciate it

**Comment by csukuangfj** [comment]
For all of you that want to export silero-vad v4 to onnx, or even to executorch, or to RKNN, please have a look at
https://github.com/snakers4/silero-vad/issues/630#issuecomment-3062488943


You can also find the PyTorch source code for silero-vad v4, which can load the weights directly from silero_vad.jit v4

**Reply by csukuangfj** [reply]
The PyTorch source code for silero vad v5 is available at
https://github.com/lovemefan/Silero-vad-pytorch/blob/main/model/silero.py
